File: Day 052/Problem 894. All Possible Full Binary Trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    unordered_map<int, vector<TreeNode*>> dp;
    vector<TreeNode*> allPossibleFBT(int n) {
        if ((n & 1) == 0) {
            return {};
        }
        
        if (n == 1) {
            return {new TreeNode()}; 
        }

        if (dp.find(n) != dp.end()) {
            return dp[n];
        }

        vector<TreeNode*> result;
        for (int i = 1; i < n; i += 2) {
            vector<TreeNode*> left = allPossibleFBT(i);
            vector<TreeNode*> right = allPossibleFBT(n - i - 1);

            for (TreeNode* leftTree : left) {
                for (TreeNode* rightTree : right) {
                    TreeNode* root = new TreeNode(0, leftTree, rightTree);
                    result.push_back(root);
                }
            }
        }

        return dp[n] = result;
    }
};

File: Day 094/Problem 894. All Possible Full Binary Trees.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    Map<Integer, List<TreeNode>> dp = new HashMap<>();
    public List<TreeNode> allPossibleFBT(int n) {
        // A full binary tree must have odd number of nodes
        if ((n & 1) == 0) return new ArrayList<>();
        if (n == 1) return List.of(new TreeNode(0));

        if (dp.containsKey(n)) {
            return dp.get(n);
        }

        List<TreeNode> result = new ArrayList<>();
        for (int i = 1; i < n; i += 2) {
            List<TreeNode> leftNodes = allPossibleFBT(i);
            List<TreeNode> rightNodes = allPossibleFBT(n - 1 - i);

            for (TreeNode leftNode : leftNodes) {
                for (TreeNode rightNode : rightNodes) {
                    result.add(new TreeNode(0, leftNode, rightNode));
                }
            }
        }

        dp.put(n, result);
        return result;
    }
}